File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    int mypaths2(int sr, int sc, int dr, int dc, vector<vector<int>> &dir,
             vector<vector<int>>&dp) {
	if (sr == dr && sc == dc) {
		return dp[sr][sc] = 1;
	}
	if (dp[sr][sc] != 0) {
		return dp[sr][sc];
	}
	int count = 0;
	for (int d = 0; d < dir.size(); d++) {
		int r = sr + dir[d][0];
		int c = sc + dir[d][1];
		if (r >= 0 && c >= 0 && r <= dr && c <= dc) {
			count += mypaths2(r, c, dr, dc, dir, dp);
		}
	}
	return dp[sr][sc] = count;
    }
    int uniquePaths(int n, int m) {
        vector<vector<int>>dir = {{0, 1}, {1, 0}};
	    vector<vector<int>>dp(n + 1, vector<int>(m + 1, 0));
        return mypaths2(0,0,n-1,m-1,dir,dp);
    }
};

File: 62-unique-paths/62-unique-paths.java
class Solution {

    public int uniquePaths(int m, int n) {
        int[][] dir = { { 1, 0 }, { 0, 1 } };
        int[][]dp=new int[m][n];
        return help(0, 0, m - 1, n - 1, dir,dp);
    }

    // public int help(int sr, int sc, int er, int ec, int[][] dir) {
    //     if (sr == er && sc == ec) return 1;
    //     int count = 0;
    //     for (int d = 0; d < dir.length; d++) {
    //         int r = sr + dir[d][0];
    //         int c = sc + dir[d][1];
    //         if (r >= 0 && c >= 0 && r <= er && c <= ec) {
    //             count += help(r, c, er, ec, dir);
    //         }
    //     }
    //     return count;
    // }

    public int help(int sr, int sc, int er, int ec, int[][] dir,int[][]dp) {
        for (int i = er; i >= 0; i--) {
            for (int j = ec; j >= 0; j--) {
                if (i == er && j == ec) 
                {
                    dp[i][j]= 1;
                    continue;
                }
                int count = 0;
                for (int d = 0; d < dir.length; d++) {
                    int r = i + dir[d][0];
                    int c = j + dir[d][1];
                    if (r >= 0 && c >= 0 && r <= er && c <= ec) {
                        count += dp[r][c];
                    }
                }
                dp[i][j]=count;
            }
        }
        return dp[0][0];
    }
}
